### PR TITLE
[RFC] constexpr for DetId

### PR DESCRIPTION
**Pull request by vkhristenko**
We discussed to put that directly to cms-sw/cmssw.

putting here for comments 100%, if all is fine... will put to cms-sw/cmssw

### CODE REVIEW

File: DataFormats/DetId/test/BuildFile.xml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<bin name="test_detid" file="test_detid.cu">
+    <use name="DataFormats/DetId"/>
```
[the remark this answers is not in the document]
Comment (vkhristenko) [reply DataFormats/DetId/test/BuildFile.xml]: true, but for some reason still worked w/o... added

[the remark this answers is not in the document]
Comment (fwyzard) [reply DataFormats/DetId/test/BuildFile.xml]: I think SCRAM may automatically pull in the dependency on CUDA when building a .cu file .
I am no 100% sure, though.

**Comment by fwyzard** [comment]
Looks fine to me...

**Comment by makortel** [comment]
+1 from me as well

**Comment by vkhristenko** [comment]
alright, I will submit a pr to main repo. thx!

**Comment by makortel** [comment]
Let me add here what @fwyzard and me just discussed.

Since the "raw to cluster" migration to HeterogeneousEDProducer requires changing `edmNew::DetSetVector` (which implies compiling ~half of the universe), we could do the following after `10_2_0_pre4` has been released
* merge this and my `edmNew::DetSetVector` changes in the patatrack branch
* make `10_2_0_pre4_Patatrack` release and install it on a couple of machines
to avoid everybody recompiling the ~half of the universe and still make progress on top of the developments.

**Comment by fwyzard** [comment]
Reference release CMSSW_10_2_0_pre3 at ...
Development branch [CMSSW_10_2_X_Patatrack](https://github.com/cms-patatrack/cmssw/tree/CMSSW_10_2_X_Patatrack) at ...
Testing PRs:
  - #55 at 61e58f664b43

### `makeTrackValidationPlots.py` plots
#### /RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/GEN-SIM-DIGI-RAW
  - [tracking validation plots](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/10824.5/index.html) for workflow 10824.5
  - [tracking validation plots](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/10824.8/index.html) for workflow 10824.8
  - [tracking validation plots](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/10824.7/index.html) for workflow 10824.7

#### /RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/GEN-SIM-DIGI-RAW
  - [tracking validation plots](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/10824.5/index.html) for workflow 10824.5
  - [tracking validation plots](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/10824.8/index.html) for workflow 10824.8
  - [tracking validation plots](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/10824.7/index.html) for workflow 10824.7

### DQM GUI plots
#### /RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/GEN-SIM-DIGI-RAW
  - reference [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-e6ba533090d6/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for reference release, workflow 10824.5
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for development release, workflow 10824.5
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_8;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for development release, workflow 10824.8
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_7;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for development release, workflow 10824.7
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for testing release, workflow 10824.5
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_8;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for testing release, workflow 10824.8
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_7;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for testing release, workflow 10824.7
  - [DQM comparison](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-e6ba533090d6/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5%3A;referenceobj2=other%3A%3A/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_5%3A;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for reference workflow 10824.5
  - [DQM comparison](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_8%3A;referenceobj2=other%3A%3A/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_8%3A;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for workflow 10824.8
  - [DQM comparison](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_7%3A;referenceobj2=other%3A%3A/RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_7%3A;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for workflow 10824.7

#### /RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/GEN-SIM-DIGI-RAW
  - reference [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-e6ba533090d6/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for reference release, workflow 10824.5
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for development release, workflow 10824.5
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_8;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for development release, workflow 10824.8
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_7;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for development release, workflow 10824.7
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for testing release, workflow 10824.5
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_8;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for testing release, workflow 10824.8
  - [DQM plots](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_7;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=none;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for testing release, workflow 10824.7
  - [DQM comparison](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-e6ba533090d6/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=other%3A%3A/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5%3A;referenceobj2=other%3A%3A/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_5%3A;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for reference workflow 10824.5
  - [DQM comparison](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=other%3A%3A/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_8%3A;referenceobj2=other%3A%3A/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_8%3A;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for workflow 10824.8
  - [DQM comparison](http://dqmgui7.cern.ch:8060/dqm/dev/start?runnr=1;dataset=/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_5;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=False;referenceobj1=other%3A%3A/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-0a4b5501feff/10824_7%3A;referenceobj2=other%3A%3A/RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1-4f368b5d1b39/10824_7%3A;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=L;root=Tracking;focus=;zoom=no;) for workflow 10824.7

### `nvprof/nvvp` profiles
#### /RelValTTbar_13/CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/GEN-SIM-DIGI-RAW
  - reference [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/reference-10824.5-step3.nvprof) for workflow 10824.5
  - development [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/development-10824.5-step3.nvprof) for workflow 10824.5
  - development [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/development-10824.8-step3.nvprof) for workflow 10824.8
  - development [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/development-10824.7-step3.nvprof) for workflow 10824.7
  - testing [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/testing-10824.5-step3.nvprof) for workflow 10824.5
  - testing [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/testing-10824.8-step3.nvprof) for workflow 10824.8
  - testing [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValTTbar_13-CMSSW_10_2_0_pre2-PU25ns_101X_upgrade2018_realistic_v7-v1/testing-10824.7-step3.nvprof) for workflow 10824.7

#### /RelValZMM_13/CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/GEN-SIM-DIGI-RAW
  - reference [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/reference-10824.5-step3.nvprof) for workflow 10824.5
  - development [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/development-10824.5-step3.nvprof) for workflow 10824.5
  - development [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/development-10824.8-step3.nvprof) for workflow 10824.8
  - development [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/development-10824.7-step3.nvprof) for workflow 10824.7
  - testing [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/testing-10824.5-step3.nvprof) for workflow 10824.5
  - testing [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/testing-10824.8-step3.nvprof) for workflow 10824.8
  - testing [profile](https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/RelValZMM_13-CMSSW_10_2_0_pre2-101X_upgrade2018_realistic_v7-v1/testing-10824.7-step3.nvprof) for workflow 10824.7

### Logs
The full log is available at https://fwyzard.web.cern.ch/fwyzard/patatrack/pulls/677084033f079dac786a97b39fba081ef32049f5/log .

**Comment by fwyzard** [comment]
@vkhristenko , these are the same changes as cms-sw/cmssw#23332, right ?

**Comment by vkhristenko** [comment]
@fwyzard same branch yep

**Pull request merged**